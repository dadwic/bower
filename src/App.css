.logo {
  display: block;
  width: 140px;
  margin-right: 48px;
}
a img.logo {
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}
.page-title {
  font-size: 64px;
  font-weight: bold;
  letter-spacing: -0.025em;
  margin: 0;
  padding: 0;
  padding-top: 30px;
  padding-left: 220px;
}
