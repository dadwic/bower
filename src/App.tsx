import React from "react";
import { Routes, Route, Navigate } from "react-router-dom";
import SearchPage from "pages/SearchPage";

export default function App() {
  return (
    <Routes>
      <Route path="search" element={<SearchPage />} />
      <Route path="*" element={<Navigate to="search" />} />
    </Routes>
  );
}
