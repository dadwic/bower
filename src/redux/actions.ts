import axios, { AxiosResponse, Canceler } from "axios";
import { AppThunk } from "./store";
import { API_PER_PAGE, API_KEY } from "../constants";
import {
  projectsLoading,
  projectsLoaded,
  projectsFailure,
} from "./projectsSlice";
import { Project } from "types/Project";
import { SortType } from "types";

const CancelToken = axios.CancelToken;
let cancel: Canceler;

export const fetchAsync =
  (text: string, sort: SortType, page: number): AppThunk =>
  async (dispatch) => {
    try {
      dispatch(projectsLoading());
      const response = await axios.get<Array<Project>>(
        `https://libraries.io/api/search?q=${text}&sort=${sort}&page=${page}&per_page=${API_PER_PAGE}&api_key=${API_KEY}`,
        {
          cancelToken: new CancelToken(function executor(c) {
            cancel = c;
          }),
        }
      );
      dispatch(
        projectsLoaded({
          data: response.data,
          total: Number(response.headers?.total),
        })
      );
    } catch (error: any) {
      if (axios.isCancel(error)) {
        console.log("Request canceled", error.message);
      } else {
        if (error.response) {
          const status = (error.response as AxiosResponse).status;
          if (status === 429) {
            return dispatch(
              projectsFailure("429 Too Many Requests! please try again")
            );
          }
          if (status >= 400 && status < 500) {
            return dispatch(projectsFailure("400 Bad Request"));
          }
          if (status >= 500) {
            return dispatch(projectsFailure("500 Internal Server Error"));
          }
        } else if (error.request) {
          console.log(error.request);
        } else {
          dispatch(projectsFailure(`Error ${error.message}`));
        }
      }
      console.log(error.config);
    }
  };
