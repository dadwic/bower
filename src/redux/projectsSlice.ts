import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Project, ProjectsState } from "types/Project";

export const initialState: ProjectsState = {
  count: 0,
  error: null,
  loading: false,
  projects: [],
};

export const projectsSlice = createSlice({
  name: "projects",
  initialState,
  reducers: {
    projectsLoading: (state) => {
      state.loading = true;
    },
    projectsLoaded: (
      state,
      action: PayloadAction<{ data: Project[]; total: number }>
    ) => {
      state.projects = action.payload.data;
      state.count = action.payload.total;
      state.loading = false;
    },
    projectsFailure: (state, action: PayloadAction<string>) => {
      state.count = 0;
      state.projects = [];
      state.loading = false;
      state.error = action.payload;
    },
  },
});

export const { projectsLoading, projectsLoaded, projectsFailure } =
  projectsSlice.actions;

export default projectsSlice.reducer;
