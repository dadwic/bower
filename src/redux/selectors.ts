import { RootState } from "./store";

export const selectLoading = (state: RootState) => state.projects.loading;
export const selectProjects = (state: RootState) => state.projects.projects;
export const selectProjectsCount = (state: RootState) => state.projects.count;
export const selectProjectsError = (state: RootState) => state.projects.error;
