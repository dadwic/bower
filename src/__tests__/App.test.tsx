import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import CssBaseline from "@mui/material/CssBaseline";
import { createTheme, ThemeProvider } from "@mui/material/styles";
import { store } from "redux/store";
import { Provider } from "react-redux";
import App from "../App";

describe("App", () => {
  it("should corretly match snapshot", () => {
    const theme = createTheme();

    const tree = renderer
      .create(
        <Provider store={store}>
          <ThemeProvider theme={theme}>
            <CssBaseline />
            <BrowserRouter>
              <App />
            </BrowserRouter>
          </ThemeProvider>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
