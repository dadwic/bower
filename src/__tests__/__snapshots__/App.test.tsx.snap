// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should corretly match snapshot 1`] = `null`;
